File: apps/ios/Domain/Tests/DomainTests/SplitsInsightsVectors.swift
import Foundation
@testable import Domain

// P1.4b: wires the real SplitsInsights.swift port into FunctionRegistry so
// splits-insights.json's vectors un-skip. Registered under
// (domain="splits-insights", fn=<name>) to match
// tools/golden-vectors/vectors/splits-insights.json exactly -- mirrors the
// Android adapter (SplitsInsightsVectors.kt) field-for-field.

private func asDebtLike(_ any: Any) -> DebtLike {
    let d = any as! [String: Any]
    return DebtLike(at: d["at"] as! String, amount: (d["amount"] as! NSNumber).int64Value)
}

private func asPaymentLike(_ any: Any) -> PaymentLike {
    let d = any as! [String: Any]
    return PaymentLike(at: d["at"] as! String, amount: (d["amount"] as! NSNumber).int64Value)
}

private func asFriendEdge(_ any: Any) -> FriendEdge {
    let d = any as! [String: Any]
    return FriendEdge(
        friendId: d["friendId"] as! String,
        groupId: d["groupId"] as! String,
        at: d["at"] as! String,
        amount: (d["amount"] as! NSNumber).int64Value
    )
}

private func asFriendSettlement(_ any: Any) -> FriendSettlement {
    let d = any as! [String: Any]
    return FriendSettlement(friendId: d["friendId"] as! String, at: d["at"] as! String, amount: (d["amount"] as! NSNumber).int64Value)
}

private func asFriendStats(_ any: Any) -> FriendStats {
    let d = any as! [String: Any]
    return FriendStats(
        friendId: d["friendId"] as! String,
        net: (d["net"] as! NSNumber).int64Value,
        youCovered: (d["youCovered"] as! NSNumber).int64Value,
        theyCovered: (d["theyCovered"] as! NSNumber).int64Value,
        lent: (d["lent"] as! NSNumber).int64Value,
        borrowed: (d["borrowed"] as! NSNumber).int64Value,
        groups: (d["groups"] as! NSNumber).intValue,
        groupsOwing: (d["groupsOwing"] as! NSNumber).intValue,
        groupsOwed: (d["groupsOwed"] as! NSNumber).intValue,
        expenses: (d["expenses"] as! NSNumber).intValue,
        avgSettleDays: (d["avgSettleDays"] as? NSNumber)?.doubleValue,
        settledDebts: (d["settledDebts"] as! NSNumber).intValue
    )
}

private func averageSettleResultToJson(_ r: AverageSettleResult) -> [String: Any] {
    ["avgDays": r.avgDays.map { jsonNumber($0) } ?? NSNull(), "clearedCount": r.clearedCount]
}

private func friendStatsToJson(_ s: FriendStats) -> [String: Any] {
    [
        "friendId": s.friendId,
        "net": s.net,
        "youCovered": s.youCovered,
        "theyCovered": s.theyCovered,
        "lent": s.lent,
        "borrowed": s.borrowed,
        "groups": s.groups,
        "groupsOwing": s.groupsOwing,
        "groupsOwed": s.groupsOwed,
        "expenses": s.expenses,
        "avgSettleDays": s.avgSettleDays.map { jsonNumber($0) } ?? NSNull(),
        "settledDebts": s.settledDebts,
    ]
}

private func friendInsightToJson(_ i: FriendInsight) -> [String: Any] {
    ["key": i.key, "friendId": i.friendId, "value": jsonNumber(i.value), "evidence": i.evidence]
}

// The same fixture edges/settlements export.ts's pickFriendInsights vector
// reuses from its own computeFriendStats vector -- that vector's
// input.stats field is just the descriptive placeholder string
// "computeFriendStats(edges,settlements) above", not real data, so this
// adapter reconstructs the identical fixture (read directly off
// tools/golden-vectors/export.ts's splits-insights section, not guessed),
// mirroring SplitsInsightsVectors.kt.
private let fixtureEdges = [
    FriendEdge(friendId: "f1", groupId: "g1", at: "2026-01-01T00:00:00Z", amount: 500),
    FriendEdge(friendId: "f1", groupId: "g1", at: "2026-01-10T00:00:00Z", amount: 300),
    FriendEdge(friendId: "f2", groupId: "g2", at: "2026-01-05T00:00:00Z", amount: -200),
]
private let fixtureSettlements = [
    FriendSettlement(friendId: "f1", at: "2026-01-15T00:00:00Z", amount: 500),
]

func registerSplitsInsightsVectors() {
    FunctionRegistry.register(domain: "splits-insights", fn: "averageSettleDays") { input in
        let d = input as! [String: Any]
        let debts = (d["debts"] as! [Any]).map(asDebtLike)
        let payments = (d["payments"] as! [Any]).map(asPaymentLike)
        return averageSettleResultToJson(averageSettleDays(debts, payments))
    }

    FunctionRegistry.register(domain: "splits-insights", fn: "computeFriendStats") { input in
        let d = input as! [String: Any]
        let edges = (d["edges"] as! [Any]).map(asFriendEdge)
        let settlements = (d["settlements"] as! [Any]).map(asFriendSettlement)
        return computeFriendStats(edges, settlements).map(friendStatsToJson)
    }

    FunctionRegistry.register(domain: "splits-insights", fn: "pickFriendInsights") { input in
        let d = input as! [String: Any]
        let statsField = d["stats"]
        let stats: [FriendStats]
        if let statsArray = statsField as? [Any] {
            stats = statsArray.map(asFriendStats)
        } else {
            stats = computeFriendStats(fixtureEdges, fixtureSettlements)
        }
        return pickFriendInsights(stats).map(friendInsightToJson)
    }
}

File: apps/android/domain/src/test/kotlin/com/sanvya/app/domain/splitsinsights/SplitsInsightsVectors.kt
package com.sanvya.app.domain.splitsinsights

import com.sanvya.app.domain.vectors.FunctionRegistry
import com.sanvya.app.domain.vectors.jsonNumber
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.double
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long

// P1.4a: wires the real SplitsInsights.kt port into FunctionRegistry so
// splits-insights.json's vectors un-skip. Registered under
// (domain="splits-insights", fn=<name>) to match
// tools/golden-vectors/vectors/splits-insights.json exactly. Every field
// here is a plain JSON number/string/bool -- this domain's exporter calls
// never wrap results with amt()/mny(), confirmed by reading the actual
// generated vector file, not assumed from earlier domains.

private const val DOMAIN = "splits-insights"

private fun JsonElement.asDebtLike(): DebtLike {
    val o = jsonObject
    return DebtLike(o.getValue("at").jsonPrimitive.content, o.getValue("amount").jsonPrimitive.long)
}

private fun JsonElement.asPaymentLike(): PaymentLike {
    val o = jsonObject
    return PaymentLike(o.getValue("at").jsonPrimitive.content, o.getValue("amount").jsonPrimitive.long)
}

private fun JsonElement.asFriendEdge(): FriendEdge {
    val o = jsonObject
    return FriendEdge(
        friendId = o.getValue("friendId").jsonPrimitive.content,
        groupId = o.getValue("groupId").jsonPrimitive.content,
        at = o.getValue("at").jsonPrimitive.content,
        amount = o.getValue("amount").jsonPrimitive.long,
    )
}

private fun JsonElement.asFriendSettlement(): FriendSettlement {
    val o = jsonObject
    return FriendSettlement(
        friendId = o.getValue("friendId").jsonPrimitive.content,
        at = o.getValue("at").jsonPrimitive.content,
        amount = o.getValue("amount").jsonPrimitive.long,
    )
}

private fun JsonElement.asFriendStats(): FriendStats {
    val o = jsonObject
    return FriendStats(
        friendId = o.getValue("friendId").jsonPrimitive.content,
        net = o.getValue("net").jsonPrimitive.long,
        youCovered = o.getValue("youCovered").jsonPrimitive.long,
        theyCovered = o.getValue("theyCovered").jsonPrimitive.long,
        lent = o.getValue("lent").jsonPrimitive.long,
        borrowed = o.getValue("borrowed").jsonPrimitive.long,
        groups = o.getValue("groups").jsonPrimitive.int,
        groupsOwing = o.getValue("groupsOwing").jsonPrimitive.int,
        groupsOwed = o.getValue("groupsOwed").jsonPrimitive.int,
        expenses = o.getValue("expenses").jsonPrimitive.int,
        avgSettleDays = o["avgSettleDays"]?.takeIf { it !is JsonNull }?.jsonPrimitive?.double,
        settledDebts = o.getValue("settledDebts").jsonPrimitive.int,
    )
}

private fun AverageSettleResult.toJson(): JsonElement = JsonObject(
    mapOf(
        "avgDays" to (avgDays?.let { jsonNumber(it) } ?: JsonNull),
        "clearedCount" to JsonPrimitive(clearedCount),
    )
)

private fun FriendStats.toJson(): JsonElement = JsonObject(
    mapOf(
        "friendId" to JsonPrimitive(friendId),
        "net" to JsonPrimitive(net),
        "youCovered" to JsonPrimitive(youCovered),
        "theyCovered" to JsonPrimitive(theyCovered),
        "lent" to JsonPrimitive(lent),
        "borrowed" to JsonPrimitive(borrowed),
        "groups" to JsonPrimitive(groups),
        "groupsOwing" to JsonPrimitive(groupsOwing),
        "groupsOwed" to JsonPrimitive(groupsOwed),
        "expenses" to JsonPrimitive(expenses),
        "avgSettleDays" to (avgSettleDays?.let { jsonNumber(it) } ?: JsonNull),
        "settledDebts" to JsonPrimitive(settledDebts),
    )
)

private fun FriendInsight.toJson(): JsonElement = JsonObject(
    mapOf(
        "key" to JsonPrimitive(key),
        "friendId" to JsonPrimitive(friendId),
        "value" to jsonNumber(value),
        "evidence" to JsonPrimitive(evidence),
    )
)

// The same fixture edges/settlements export.ts's pickFriendInsights vector
// reuses from its own computeFriendStats vector -- that vector's
// input.stats field is just the descriptive placeholder string
// "computeFriendStats(edges,settlements) above", not real data, so this
// adapter reconstructs the identical fixture rather than trying to parse
// it (read directly off tools/golden-vectors/export.ts's splits-insights
// section, not guessed).
private val FIXTURE_EDGES = listOf(
    FriendEdge("f1", "g1", "2026-01-01T00:00:00Z", 500L),
    FriendEdge("f1", "g1", "2026-01-10T00:00:00Z", 300L),
    FriendEdge("f2", "g2", "2026-01-05T00:00:00Z", -200L),
)
private val FIXTURE_SETTLEMENTS = listOf(
    FriendSettlement("f1", "2026-01-15T00:00:00Z", 500L),
)

fun registerSplitsInsightsVectors() {
    FunctionRegistry.register(DOMAIN, "averageSettleDays") { input ->
        val o = input.jsonObject
        val debts = o.getValue("debts").jsonArray.map { it.asDebtLike() }
        val payments = o.getValue("payments").jsonArray.map { it.asPaymentLike() }
        averageSettleDays(debts, payments).toJson()
    }

    FunctionRegistry.register(DOMAIN, "computeFriendStats") { input ->
        val o = input.jsonObject
        val edges = o.getValue("edges").jsonArray.map { it.asFriendEdge() }
        val settlements = o.getValue("settlements").jsonArray.map { it.asFriendSettlement() }
        JsonArray(computeFriendStats(edges, settlements).map { it.toJson() })
    }

    FunctionRegistry.register(DOMAIN, "pickFriendInsights") { input ->
        val statsField = input.jsonObject.getValue("stats")
        val stats = if (statsField is JsonArray) {
            statsField.map { it.asFriendStats() }
        } else {
            computeFriendStats(FIXTURE_EDGES, FIXTURE_SETTLEMENTS)
        }
        JsonArray(pickFriendInsights(stats).map { it.toJson() })
    }
}
